Scheduler: Use a "scheduler" task for thread sleep

kpamnany opened the pull request:
A Julia thread runs Julia's scheduler in the context of the switching task. If no task is found to switch to, the thread will sleep while holding onto the (possibly completed) task, preventing the task from being garbage collected. This recent [Discourse post](https://discourse.julialang.org/t/weird-behaviour-of-gc-with-multithreaded-array-access/125433) illustrates precisely this problem.

A solution to this would be for an idle Julia thread to switch to a "scheduler" task, thereby freeing the old task.

This PR uses `OncePerThread` to create a "scheduler" task (that does nothing but run `wait()` in a loop) and switches to that task when the thread finds itself idle.

Other approaches considered and discarded in favor of this one: https://github.com/JuliaLang/julia/pull/57465 and https://github.com/JuliaLang/julia/pull/57543.

kpamnany committed:
Use a scheduler task for thread sleep

vtjnash commented on base/task.jl:
```diff
@@ -1184,7 +1196,12 @@ checktaskempty = Partr.multiq_check_empty
 @noinline function poptask(W::StickyWorkqueue)
     task = trypoptask(W)
```
I think this should fix your concern that we don't do enough of the `jl_task_get_next` loop before switching to the scheduler to spin and wait (though we could also move more of that loop here)
```suggestion
    task = trypoptask(W)
    if !(task isa Task)
        process_events()
        task = trypoptask(W)
    end
```

kpamnany commented on base/task.jl:
```diff
@@ -1184,7 +1194,16 @@ checktaskempty = Partr.multiq_check_empty
+        if current_task() === sched_task
+            task = ccall(:jl_task_get_next, Ref{Task}, (Any, Any, Any), trypoptask, W, checktaskempty)
+        else
+            yieldto(sched_task)
```
A `@goto` the beginning of `poptask` here resulted in a freeze while building:
```
Random  ────────  1.407750 seconds
Stdlibs total  ─ 20.005079 seconds
Sysimage built. Summary:
Base ────────  58.894729 seconds 74.6422%
Stdlibs ─────  20.005079 seconds 25.3541%
Total ───────  78.902671 seconds
^C
```
🤔

vtjnash replied:
Yes, if the first thing a Task does after getting scheduled is to go back to sleep, nothing will ever get done

vtjnash replied:
You are missing an assignment of task=current_task() here

kpamnany committed:
Rearrange the code to make things cleaner

vtjnash commented on base/task.jl:
```diff
@@ -1191,35 +1191,44 @@ end
+        # didn't find a task to run, try again
         process_events()
         task = trypoptask(W)
     end
```
this seems unneeded now (not bad, just unnecessary)

kpamnany replied:
Will remove if I "feel" no added latency at the REPL.

kpamnany replied:
Removed.

kpamnany commented:
This is currently blocked on what seems to be a bug in `OncePerThread` serialization/deserialization; found with @gbaraldi.

kpamnany commented:
Unblocked... thanks @gbaraldi!

Now, some `Channel` tests and a `Sockets` test are failing. Looking into these failures.

vtjnash reviewed (approved):
I think those tests are specifically testing race conditions in the code, so you might need to adjust them slightly to account for the change in `process_events` ordering and count

vtjnash commented on base/task.jl:
```diff
@@ -1181,25 +1191,44 @@ end
+    have_result = false
+    W = workqueue_for(Threads.threadid())
+    task = trypoptask(W)
+    if !(task isa Task)
+        # didn't find a task to run, try again
+        process_events()
+        task = trypoptask(W)
+    end
```
```suggestion
```
I don't think you need to call `epoll` twice in this rapid succession

JamesWrigley commented:
Is there any chance of getting this backported to 1.12? It's quite tricky to debug and would be nice if it was fixed in the next release.

kpamnany committed:
Drop extra process_events call

kpamnany committed:
Fix channels tests

This small group of tests is written with assumptions about when
and how the libuv event loop is run. As this PR changes this
behavior, the tests needed adjusting.

kpamnany commented:
The `channels` tests are fixed, but I don't see a way to fix the `Sockets` test that's failing. Any ideas @gbaraldi or @vtjnash?

kpamnany committed:
Fix Sockets test

Previously, this test depended on scheduler behavior, which is
slightly changed in this PR. Changed the test to connect to a
non-routable IP address so that it no longer depends on task
ordering.

kpamnany commented:
The `channels` tests that are failing on FreeBSD are a bit mystifying. How come `WorkQueue` is empty on Linux but not on FreeBSD? Do we have an extra sticky task?

kpamnany committed:
Remove bad tests for `Workqueue` contents

nsajko commented on base/task.jl:
```diff
@@ -1181,25 +1191,39 @@ end
+    # We may have already switched tasks (via the scheduler task), so
+    # only switch if we haven't.
+    if !have_result
+        @assert task isa Task
```
Is there any specific reason this doesn't just typeassert?

```suggestion
        task = task::Task
```

kpamnany replied:
No specific reason. Seems like it should be a typeassert, actually.

Keno commented:
This change interacts badly with ^C. Many interrupt exception will now get thrown to the scheduler task, which then dies and takes down the ability to ever schedule again in the future.

kpamnany commented:
We could also simply never direct the interrupt at scheduler tasks?
